Improvements to AudioLibrary.GetArtists, GetAlbums and GetSongs

DaveTBlake opened the pull request:
API improvements to AudioLibrary.GetArtists, GetAlbums and GetSongs. This follows on from  #8015 and the addition of artist roles to the music library, and adresses some flaws that were in the initial API implementation for accessing artist role data.

It also attempts to fix some longer standing issues: GetArtists now returns the song genres and genreids of the songs of each artist, and a flag if the artist is an "album artist". Thus returning the data values that correspond to the filters. It  makes it clearer in the documentation that the only genre filtering currently offered is song genre (not the artist genre value that can be scraped, stored and displayed but is not used in queries). Extending filtering to all kinds of genre (artist and album) is something for another PR.

@Montellese you have seen all of this before, so hope that all is well.
@JoeTheFox finally catching up with the iOS remote!
@Tolriq I think this covers the many things you pointed out with roles, but also improves GetArtists generally in the way I said I would. More work to do of course, especially with genre, but this seems enough for one PR.

DaveTBlake committed:
API improvements GetArtists, GetAlbums and GetSongs for accessing artist role data.
Return roleid as part of SongDetails.Contributors.
Make artistid/artist, roleid/role a compulsory pair in filter for GetAlbums and GetSongs.
Document filtering highlight song genre used for all genre filters.
Note that for backwards compatibility there is implicit filtering to just artists, excluding other contributor roles. Use parameter flag in GetArtists,for role = all to return all contributors, and in GetAlbums and GetSongs so all roles allowed when filter by artist.

AudioLibrary.GetArtists returns roles and roleids

Return song genre and genreid in GetArtists

Return isalbumartist field from GetArtists indicating those album artists

GetArtists returns songgenre and role as array of (id, value) pairs.
Correct Audio.Details.Artist missing fields.

Remove preloading of composer and conductor roles

Tolriq commented on xbmc/interfaces/json-rpc/AudioLibrary.cpp:
```diff
@@ -51,16 +51,28 @@ JSONRPC_STATUS CAudioLibrary::GetArtists(const std::string &method, ITransportLa
+    genreID = (int)filter["genreid"].asInteger();
+  else if (filter.isMember("genre"))
+    musicUrl.AddOption("genre", filter["genre"].asString());
+  if (filter.isMember("songgenreid"))
```
else if ?

But if genre / genreid are deprecated in that case, songgenreXX should be tested first to prevail ?

DaveTBlake replied:
As written songenre/songenreid will prevail over genre/genreid. The later uses song genre any way, as it always has, so left operational to give API consumers a chance to change

Tolriq replied:
Yes but previously all was a big else if block.

Since now it's multiple block the if are hidden in the middle of the else if and this make the code not readable. And error prone in the future if new fields added.
At least add empty lines to separate the block and avoid the confusion.

The remark also covers that if you add the 2 fields in JSON query AddOption will be called twice, leading from just this code reading to unexpected behavior. 

And allowing both fields to actually do something is against your comment just before that says only 1 allowed.

DaveTBlake replied:
Just spotted this comment, and of course I can add the spaces but

>The remark also covers that if you add the 2 fields in JSON query AddOption will be called twice, leading from just this code reading to unexpected behavior. 

sorry I don't understand what you are saying here?

>And allowing both fields to actually do something is against your comment just before that says only 1 allowed.

The restriction on only one filter value (of fooid type) allowed is applied by the definition in methods.json, it does not rely on the elseif here.

Tolriq commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -641,26 +641,35 @@
           { "type": "object", "properties": { "songid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
+          { "type": "object", "properties": { "songid": { "$ref": "Library.Id", "required": true }, "roleid": { "$ref": "Library.Id", "required": true }}, "additionalProperties": false },
+          { "type": "object", "properties": { "songid": { "$ref": "Library.Id", "required": true }, "role": { "type": "string", "minLength": 1, "required": true }}, "additionalProperties": false },
           { "type": "object", "properties": { "roleid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
```
Remark about all roles applies to roleid too, so add comment too ?

Tolriq commented:
Thanks for this PR.

A few comments in code from quick review.

Only important remaining thing is how to warn API consumers that there's breaking changes here.

@Montellese validated that we should not bump major revision 10 times during this v17 that will have a few ones. 

Major will be bumped to 8 at RC.

But since the PR message is lost during merge, I have no idea how to leave something easy to see about the breaking change :(

Don't know if we can put comments in version.txt maybe @Montellese have ideas.

DaveTBlake commented:
>@Montellese validated that we should not bump major revision 10 times during this v17 that will have a few ones. 

Not sure what you are saying here @Tolriq

Tolriq commented:
@DaveTBlake you are making breaking changes here so the API version should be 8.X.X.

But due to what I said previously, it have been decided that during v17 alpha the major version will not be bumped multiple times to avoid absurd API numbers.

API will be set at 8 during RC.

But now the question is how do we leave an easy way for users to see that there's a breaking change.

DaveTBlake commented:
>But now the question is how do we leave an easy way for users to see that there's a breaking change.

No idea. But not sure how much of an issue that is during the alpha phase of Krypton? Released API needs to show a breaking change, hence "8", but during development aren't such changes to be more expected? For skins (that do not have the luxury of a core interface version number) there is a forum thread where the impact of every interface change is highlighted.

Anyway is it actually not still backwards compatible? It does not remove "genre" as a GetArtists filter, just adds "songgenre" and comments to use it. By default it only gives role=1 artists, the same results are before.  It may deserve a major version change to highlight the new facilities, but this PR won't practically stop things from working.

There are a couple of API version changing PRs waiting to merge at the moment so the minor number may need to be set just before merging anyway depending on what merges first.

Tolriq commented:
Your PR rename functions in JSON that exist in API 7.x, so it's a breaking change.

The impact is always the same, if you break compatibility, there must be a way for consumers to know.

Else how do you deal with product that want to support Kodi alpha due to user asking, but can be broken every night without a way to detect ?

The major version change is the indicator, a remote that support API 7.X is ensured that as long as Kodi return 7.X it will still work. If Kodi start to return 8.X then the remote can inform the user that this version is not supported.

If not then the remote no more works, the users complains and it's bad for everyone.

But then the dev tried to find out why it's broken and have no way to find there was a breaking change since major version is not changed.

So yes we can say, it's alpha it will break do not use ...

But then everyone wait for beta / rc to start coding for support ? Then discover broken things and it's too late to fix in core level ?

From day 1 the versioning scheme was respected and is important for the global eco system around Kodi.

So IMO breaking change during alpha without major bump implies that there's a way for API consumer to easily find out.

There was a thread http://forum.kodi.tv/showthread.php?tid=98551 but unmaintained as the version.txt file is the correct and easy way to see compatible and incompatible changes :)

As I said maybe it's possible to add a comment in version .txt to give those breaking change a trace and it could be OK.

joethefox commented:
Sorry @Tolriq , I'm using this patch from a while and didn't notice any backward compatibility issue, can you point me the exactly breaking things so I can double check my code? Which functions were renamed?

Tolriq commented:
Well easy to see in Code ;)

Function rename :
https://github.com/xbmc/xbmc/pull/9300/files#diff-c5ebd35927606ffb9137e86cb28a4120L114

Positional parameter changed : 
https://github.com/xbmc/xbmc/pull/9300/files#diff-1bb266d43e6d1a19799e53fcf6371260L466

joethefox commented:
GetContributorRoles vs GetRoles? are you serious? GetContributorRoles was introduced during this 17 alpha stage so it's a non issue. About the positional parameter changes, how this can be a backward compatibility issue?

Tolriq commented:
Read all the messages from @Montellese about that maybe ?

About non issue, maybe you should read my comment, it's obviously an issue. Or remotes should not use any 7.x API because they can break without a way to detect ? So users do not use alpha ? So no one test things before release ? This is a very wrong approach to the problem.
Breaking changes during alpha or at any time must be documented and easy to see by any API consumer.

There's the correct way to manage things and the bad way where's shortcuts are taken.

I never said this is a blocking issue, I said that API consumers need to know it's a breaking change, because yes it is.

I'll let @Montellese explain and decide, this is tiring to try to protect the integrity of Kodi APIs and would not have expected this from you @joethefox .

What if one of the new function you use is renamed in 10 days without bump ? Your remote no more works, you need to fix, to ask all users to update or revert Kodi, then wait for a validation from Apple ?
All that without an easy way to find out what did break your code ?

The alphas are public and easy to download by end user, so they should be handled correctly.

joethefox commented:
Again apart GetContributorRoles vs GetRoles, that nobody (apart you) can consider an issue, can you point me the exact remaining issues so I can test the remote app? With or without this app the code I wrote SEEMS working, but if you are so kind to list me the other breaking functions then I can understand you, otherwise I'm not understanding anything of what you say, my sincere apologies

Tolriq commented:
Ok so because you already are using the new naming and not the previous one you do not see a breaking change ?

But if you had used the previous name this change would be a breaking one no ?

I'm sorry but I do not understand any of your argument about this not being a breaking change.

I only read, my code is not impacted since I already use this new naming, so there's no problem.

DaveTBlake commented:
@joethefox I think @Tolriq is talking about general principles, while you (and I) are talking about this particular instance.

@Tolriq I am happy to number this however is required, and I understand the principles and why they are important. But there is a clear conflict between using the numbering system, and also saying that major version change will only happen with RC. Clearly we need to make backwards incompatible changes before RC so they can be tested, and they may well happen in separate PRs too.

@Montellese how do you want to handle this? Can I suggest take a similar approach to OpenElec version numbering?

Tolriq commented:
@DaveTBlake yes I perfectly know the conflict problem that's why since my first post I said we need to find a solution and pinged @Montellese about this ;)

In that case the numbering problem was decided, it just miss the solution to warn API users during the alpha stages.
@Montellese is very busy so it was hard to have a decision and a complete discussion about this, meaning this last part was not yet decided, but this is something that we have started discussing since the 7.x bump in multiple PR.

I don't know why such answers from you both on my comment that was targeted at  @Montellese from the start about a problem that @Montellese is aware since the start and that he needs to decide.
All about things that I follow since a long time so that I know questions and consequences ;)

DaveTBlake commented:
>I don't know why such answers from you both on my comment that was targeted at @Montellese ...

Well as a comment on this PR I take the content as being specifically aimed at me and this work. I don't think that this PR is the right place for a more general discussion about API version numbering and change management.

>@Montellese is very busy so it was hard to have a decision and a complete discussion about this...

Yes that is both very true, and frustrating. Meanwhile what do I do? This work was done in January, Joe has already released to use it, the change is needed and improves things. But we don't know what to do with version numbering so it just stalls?

Tolriq commented:
Well you are unlucky but you are the PR that triggers the mandatory of the discussion, since you are the PR that brings a  breaking change after the decision to not bump major version number during alphas.

So I'm sorry that we both are frustrated by this coming again on you and depending on someone else for resolution ;) But AFAIC I have to continue my work that some do not like to help Kodi keep stable and working API.

So IMO either it stalls until @Montellese can find a solution that suits, or we keep previous working way and API becomes 8.X.

This is as always not my call, but I hope that integrity prevails about merging speed, as we all know that once something is in, it's always the reason for the next and the next and the next thing to follow as it was done once.

joethefox commented:
allow me to disagree, you are talking about a name change during the same alpha stage and apart this nothing else is changed. So, meanwhile your general thought is right, it can't be applied to this PR.

Tolriq commented:
Read all the Montellese discussion about that maybe as I said ?

If alpha change are not relevant then why bother updating minor and revision fields ? 
They means nothing after all no, it's only alphas ?
Why notifying that there's a fix, there's tons of fix during alphas .....
Why notifying that there's new functions, it's alphas there's always new functions ....

Let's make API version the Kodi version, why bother no ? 17.0.1 for alpha 1 then continue ?

This is an API the release are public since distributed on Kodi website, so API is public, so it needs proper management.

I'll continue to disagree there's no valid reason to not consider alpha as not normal release.

Rules are rules, exceptions are anarchy in the case of public API.

joethefox commented:
Alpha changes are relevant but during this alpha stage was introduced a method named GetContributorRoles and after a couple of weeks was renamed to GetRoles, nothing else has changed. 

Your thought is already well explained, which I agree, but please or you are so kind to point me the other actual breaking things or it seems a little bit out of place if applied to this PR.

Tolriq commented:
Well a little history : 
https://github.com/xbmc/xbmc/pull/8268
https://github.com/xbmc/xbmc/pull/8306

Just a few days between, all in the same alpha, all minors but yet 2 bump because this is the way to go.

This is not out of place, this is a breaking change, I have tons of example of API bump and why it's necessary and the fact that @Montellese always pushed for that and agreed.

On the other side I do not see any reason to not consider this a breaking change, I only see this is an exception, but no it's not. There's no valid reason to create this as an exception.

Since there's no argument or rules that would describe this as an exception, this means that any breaking change during any alpha does not need to bump API version.
This means API version is useless, this means all the work made since day 1 with @Montellese to secure API management would be completely lost.

Show me a rule that can apply to this case and will not open the door to future breaking API change during alphas and I'll gladly consider it.

But ATM I see none and you have given none.

joethefox commented:
To better understand, looking at this PR you are talking about GetContributorRoles vs GetRoles or is there else more?

Montellese commented:
First of all renaming a method (even if it was introduced a few days/hours ago) is breaking backwards compatibility. So it requires a version bump.
Adding a property to an object is NOT breaking backwards compatibility because object properties an NOT be used in a numbered order. That only applies to parameters. But inside a parameter (which can be an object) it's not an issue. We tend to add new properties, enum values etc. at the end because it makes reading the change easier but it's not an issue for the API.

Concerning the version bumping I think we should stay in the 7.x.y range and bump to 8.0.0 when branching for (beta) release. At that point it's not allowed anymore to change the API anyway, only fix implementation details (which can bump the fix number). So the release will be with 8.0.x. I admit that this is not ideal for JSON-RPC client developers because they can't differentiate between a non-breaking and a breaking change but otherwise we will release Krypton with a JSON-RPC API version of 27...

Tolriq commented:
@Montellese 
> I admit that this is not ideal for JSON-RPC client developers because they can't differentiate between a non-breaking and a breaking

This was the first question here :( What can we add so that clients can easily find the breaking changes. (I also pushed toward the no API 27 during alpha in our previous exchanges and using odd / even number scheme as Linux kernel)

Adding a comment in version.txt ? Adding a new txt file ? Anything that a dev can refer too, to see each breaking change with each version number.

Without anything, it's really hard for anyone to find the culprit PR and associated version and branch something in it's code.

We can also maybe use the previous thread you managed but I doubt you have the necessary time for that :(

We just need to find a way to track those, so that devs have a reference to follow and are not lost trying to understand and find the PR's that broke something. Git log for version.txt is enough for all changes, but does not include the breaking type thing.

DaveTBlake commented:
>Git log for version.txt is enough for all changes, but does not include the breaking type thing.

Is the PR comment in Git not enough? The version number changes, look at the PR that made that change and read what it says?

Tolriq commented:
Well no as the version update is marked as backward compatible, dev have no reason to find this specific PR as the culprit.

Imagine I have tested Yatse with API 7.1.0. 
7.3.0 introduce a breaking change, but no user report before someone test with API 7.9.0 (imagine there's a lot of JSON change). How do I find that the problem is 7.3.0 ?

Devs should not have to read potentially 20 or more long commits messages to find the one that cause problem.

But this makes me thought of a solution.

IMO this would be enough (if @Montellese  agree), when there's a breaking change, make the version.txt change a distinct commit in the PR and prefix this commit title with something like [BREAKING CHANGE].

This is simple enough and allows devs to quickly find breaking changes to JSON API.

DaveTBlake committed:
[Breaking Change] AudioLibrary.GetContributorRoles renamed AudioLibrary.GetRoles

Montellese commented:
That works for me. I usually also prefix all commits with JSON-RPC changes with `JSON-RPC:` or `[JSON-RPC]`. That also makes it easier to find the commits.

DaveTBlake commented:
What can I do to move this forwards?
@Montellese are you happy? We have discussed most of it already.
@Tolriq would you like a test build?

Tolriq commented:
@DaveTBlake can't test before Monday but yes a test build would spare me a lot of minutes of full rebuild.

Only thing maybe would be to use this PR to also add songgenreid and songgenre filter, deprecate the previous one and return songgenres for albums too to harmonize the API and prepare a maybe future filter on scraped album genre but not mandatory at all.

DaveTBlake commented:
I really want to get this sorted before the merge window closes end of this week, but obviously need you approveal @Montellese . Any issues?

@Tolriq a Win32 build will be on the mirrors shortly.

Despite the version number issues of yet another change, I would rather leave changing GetAlbums filter to  songgenreid and songgenre to another PR. The situation with albums is different from than that of artists, which was just about naming clearly what already happened, and needs deeper changes. Genre in Kodi - artist, album and song - is due a refactoring, which from recent experience I expect to be a long and painful job. I need a break before I take it on. Meanwhile the improvements in this PR are worth release on their own.

Tolriq commented:
@DaveTBlake have tested most things and all works correctly and finally allows everything users asked about artists genres :)

Just for sanity check can you point me the necessary TAGS to add non standard roles ? My collection only contains Composers and can't find how to add more TAGS in musicbrainz.

DaveTBlake commented:
>Just for sanity check can you point me the necessary TAGS to add non standard roles ? My collection only contains Composers and can't find how to add more TAGS in musicbrainz.

@Tolriq it depends what you are tagging, but Picard often populates a variety of role data, lyricists, producers, instrumentalists etc. This table may help http://kodi.wiki/view/Music_tagging#Tags_Kodi_reads generally.The non-standard tags go in the TMCL (mp3) or Performer (FLAC) tags, Picard displays it slightly different to how it is stored e.g. Billy Wizz (drums)

Thanks for testing.

Tolriq commented:
@DaveTBlake seems there's some problems with music DB, I have some problems to reproduce always but 2 times when I tried to remove a source and cleanup the associated data it did not delete anything in DB.

Logs contains errors : 
```
11:15:01 T:20136   ERROR: SQL: [MyMusic60.db] SQL error or missing database
                                            Query: INSERT INTO tmp_delartists VALUES(1)')
11:15:01 T:20136   ERROR: Exception in CMusicDatabase::CleanupArtists() or was aborted
```

Starting from scratch I then discover that Estuary does not correctly display Roles in the Roles node. (It's now always empty, before there was the Composer node)
(Db does contains 6 roles)

I also have big problems to add correct TAGS.
Despite all my effort I was unable to import TMCL tags, MP3TAG confirm that the tags are 2.4 (Took me a while to see the difference). But they are never imported.

Musicbrainz have added a IPLS tag before I convert to 2.4 and this IPLS tag is correctly read.
I can add new values there.

I lost quite a few hairs and hours in all this process. I suppose users will have hard times with this too.

Now about JSON :

JSON GetRoles returns 4 roles while there's 6 in the DB. (Composer and Conductor roles are not returned since not used in this last 1 album test).

I know we will probably disagree, but API should return all data in Kodi.
If by default you want to hide currently unused roles for easier usage from GUI for example, please add an allroles options like other to return all the data.

DaveTBlake commented:
Thanks for testing feedback @Tolriq 
cleanupArtists: I can see the  error and will fix. Good catch.

Tagging: TMCL, TIPL are v2.4 tags, and varied support by tagging software I guess.  I have used Mp3tag and TMCL tags so not sure what went wrong for you. But I don't see creating tags as a Kodi problem, just that Kodi processes them if there correctly (need to be pairs as ID3 standard specifies). I will have another look, and we probaby need a wiki with user advice for this new feature, but not part of this PR.

And to JSON

>I know we will probably disagree, but API should return all data in Kodi.

Mostly when we have disagreed it is when I am attempting to be consisent with the rest of Kodi (and API), but the rest of Kodi isn't doing what you think it should. 

>If by default you want to hide currently unused roles for easier usage from GUI for example, please add an allroles options like other to return all the data.

Sure I will do. But do you realise that GetArtists does not return the "unsued" artists, it never has. It always checks that there are song_artist and/or album_artist entries, orphaned artists left after music files removed/changed but before clean up done, are not returned.

Tolriq commented:
Well for tags I report the problem I had to be able to test this :)

I suppose it's not possible to have 2.3 and 2.4 at the same time in a mp3 file. And MP3Tag did report the tags to be 2.4 but contains the old IPLS tag. 
Maybe it was an MP3tag problem during the conversion that forget to change the IPLS tag to TMCL, but since from all the link you gave IPLS in 2.4 should not work I think there's something to address to be consistent with theory.

I suppose it should be easy to reproduce by adding IPLS tag to a 2.4 file and see what Kodi do.

For estuary after adding some more songs, it now shows a Lyricist role. So maybe the problem is on data the skin get and not the skin. (Maybe tied to compilation artists, but the artists that are on the unseen roles are visible in the artist view so unsure)

About our disagreements I suppose it's not the place to talk again, but the fact that Kodi history does something does not always means it does the correct things. Specially on music that was often forgotten in Kodi work.

DaveTBlake commented:
>If by default you want to hide currently unused roles for easier usage from GUI for example, please add an allroles options like other to return all the data.

I think that excluding unused roles by default (as currently) makes practical sense. And I am happy to add an option to also return the unused roles to the API. But if I simply do that it raises 2 concerns: 

a) It will mean that GetRoles is returning filtered data without returning the value it is filtered on, or I also add a "unused" flag that also gets returned (and it feels the issue gets bigger and bigger)

b) The principle that no filters means all data is contradicted. This is the  case with GetArtists because there was no reasonable backward compatible way around it, it just broke too much. Are you, and @Montellese, happy for GetRoles to do that?

The other approach is just to abandon the idea of predefined/pre-populated roles - all roles (except 1) are used or they get deleted. My addition of composer and conductor role on table creation was accidental, it helped with initial testing but cleaning (if that currently worked) will remove any unused. I am undecided if holding standard roles with set IDs has a use. Maybe some kinds of UI would benefit? My early ideas all roles were fixed, maybe ths is just a hangover from that? Either way it either needs implementing fully, or removing.

All points on your previous post noted @Tolriq, I just have no comment to make at this time.

Tolriq commented:
I agree on all, but tried to find a simple solution to avoid you saying I'm blocking :p

IMO unused roles should be deleted on cleanup. As are artists. 
And all roles returned before cleanup if there's unused.

I was not aware of implicit filter on artists when no songs are available since I always cleanup library after scan / change. 
But again IMO this should not be the case, the user have chosen to not cleanup but we hide the uncleaned things both to the user and the API.

There's should not be hidden things, if the user click on a artist that then have no song or albums the GUI should show a message informing the user that he should cleanup it's library.

There's some kind of similar problems with Tv Shows, should they be removed or not when no more episodes but still root folder and things like that (leading to hideemptyseries advsetting for example).

This is not a simple problem.

DaveTBlake commented:
Because so many areas behave that way, hiding orphaned data before it is cleared, it seems like a design principle rather than a mistake. In your view they are wrong, and that may be the common view now, personally I am not bothered either way and happy to accept the historic design. I certainly don't have the time or energy to work though all the areas and consistently change it. But I do believe that kind of change, if it is to change, needs to be made consistently.  

The only discussion pertinent to this PR is what happens with roles. The simple solution is to correct the initial addition of composer and conductor roles. If we do later discover that having some standard roles, with fixed and well known IDs, is useful then it can be added and the API extended correspondingly. 

Having done that do we want to hold back all the good improvements in this PR because, like other areas, it hides orphaned data prior to cleanup?

I am not saying never return orphaned data, I am suggesting that the design principle of returning orphaned data is something that is applied across the board in a separate PR.

Tolriq commented:
About roles there's the same problem as missing artist and translations, in theory, this needs a global approach about runtime translations for those things. But since the role list can be infinite this becomes hard to manage with fixed ID in all cases.

Anyway I never block anything. (Nor have the power too ;) )
The only attempt to block something that was broken and against some JSON RPC principles ended up in the way you know ;)

I just give my opinion on things that from my point of view are not optimal to fully open Kodi to the outside world and unleash it's full potential.

DaveTBlake commented:
@Tolriq you do seem to be an infinite fount of Kodi problems! That is useful up to a point, but when it gets poured into latest open PR discussion it gets overwhelming. I know it seems like there is nowhere else for observations to go, but I'm not convinced Git is the place either.  You give your opinion, well do remember what Dirty Harry said about opinions. :)

Translating roles into user language.... Possible for the standard roles composer, conductor, lyricist, remixer etc. and then for instruments etc. it is down to the user how they tag e.g. violin, geige or violon, vocalist or chanteur. But let's walk before we run.

But back with this PR. I think you were thrown by seeing 6 records in the role table and then none on the roles node in Estuary. There are 2 things happening here: a) that build did preload roles that could have been unsed, and b) unlike genre node where all used song genres get shown only the standard roles that are used appear by default. So if you had a "violinist" role from tags to see it needs a custom node creating. This is just how I introduced the feature to the UI. A chicken and egg situation between database/core features and UI. A roles node more like genre can be created, it just hasn't been done yet. But roles comes into its own as a filter when combined with genre, and that is something that takes custom nodes to cater best to the user's particular interests.

You give opinions, I give explanations :)

You see it as suboptimal that the API does not return orphaned artists, nor orphaned roles. I say Ok, that could be true but it is not a crisis. It is a general aspect of the API that is best handled and changed consistently in a separate PR. When I say something to be done separately later I generally do come back and do it.

DaveTBlake commented:
@Montellese may I merge this now?

To cut through the debate with @Tolriq: 
I have rebased to fix the library clean up problem @Tolriq  found in testing - a SQL typo fixed in #9361 

The UI is behaving as expected, even though I note it can be improved along with adding wiki user help on tagging.

The preloading of composer and conductor roles has been removed (left in accidentally/incomplete feature).

Having benefitted from testing feedback, I would like to note that orphaned roles, like orphaned artists, are not returned to the API and list that issue for separate work at a later date.

Montellese commented:
I'll try to find the time to review it tomorrow.

Montellese commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -651,26 +651,35 @@
-          { "type": "object", "properties": { "genre": { "type": "string", "minLength": 1, "required": true } }, "additionalProperties": false },
+          { "type": "object", "properties": { "genreid": { "$ref": "Library.Id", "required": true, "description": "Deprecated, use songgenreid. Filter for existance of songs with this genre" } }, "additionalProperties": false },
+          { "type": "object", "properties": { "songgenreid": { "$ref": "Library.Id", "required": true, "description": "Song genreid. Filter for existance of songs with this genre" } }, "additionalProperties": false },
+          { "type": "object", "properties": { "songgenreid": { "$ref": "Library.Id", "required": true }, "roleid": { "$ref": "Library.Id", "required": true }}, "additionalProperties": false },
```
Could you use separate lines here for the two properties so that it's easier to see? First I was wondering why there are three identical `songgenreid` filters.

DaveTBlake replied:
I will apply separate lines to all the combined properties I have added

Montellese commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -651,26 +651,35 @@
         ]
-      }
+      },
+      { "name": "allroles", "type": "boolean", "description": "Whether or not to only include all artists irrespective of the role they contributed. If the parameter is not passed or is passed as null then it is taken as false. When true it overrides any role filter value." }
```
You need to adjust the description. It can't be `null` only `true` or `false`. And you are missing a default value.

DaveTBlake replied:
With allroles I was attempting to mimic how the albumartistsonly parameter is specified.

Should this be 
{ "name": "allroles", "type": "boolean", "required": false, "description": "Whether or not to only include all artists irrespective of the role they contributed. The parameter defaults to false. When true it overrides any role filter value." }

or

{ "name": "allroles", "$ref": "Optional.Boolean", "description": "Whether or not to only include all artists irrespective of the role they contributed. The parameter is not passed or is passed as null then it is taken as false. When true it overrides any role filter value."}

or something else?

Montellese replied:
You cannot compare this parameter to `albumartistsonly` because `albumartistsonly` has three states (`true`, `false` and `whatever the GUI setting says`) whereas `allroles` only has two states (`true` and `false`), so making it an `Optional.Boolean` doesn't make any sense. Furthermore any parameter/propertiy without an explicit `"required": true` is always optional and every optional parameter needs a `"default": ...` definition. So what you are missing is `"default": false` and you need to adjust the description because it can't be `null`.

DaveTBlake replied:
So like this:
{ "name": "allroles", "type": "boolean", "default": false, "description": "Whether or not to only include all artists irrespective of the role they contributed. When true it overrides any role filter value." }

Montellese replied:
Yes.

Montellese commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -460,14 +462,37 @@
+    "items": { "type": "object",
+      "description": "The various genres of one or more songs by an artist",
+      "properties": {
+        "songgenreid": { "$ref": "Library.Id", "required": true },
```
I'm not sure if we should repeat the `songgenre` part here. The property is already called `songgenres` so repeating `songgenre` in the properties inside that property feels redundant. And if we name the properties `id` and `genre` (or `name`) they can be reused for other genre properties.

DaveTBlake replied:
Could create a generic genre id, name pair for general (audio?) use, but then would need to call it something other than `Audio.Artist.SongGenres`. Advice on naming if you want that?

Montellese replied:
`Audio.Details.Genre`?

Montellese commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -176,8 +176,6 @@ void CMusicDatabase::CreateTables()
   CLog::Log(LOGINFO, "create role table");
   m_pDS->exec("CREATE TABLE role (idRole integer primary key, strRole text)");
   m_pDS->exec("INSERT INTO role(idRole, strRole) VALUES (1, 'Artist')");   //Default role
-  m_pDS->exec("INSERT INTO role(idRole, strRole) VALUES (2, 'Composer')");
```
Do we need extra logic to remove those from databases where they aren't needed (anymore) or do we assume that they go away during a library cleaning?

DaveTBlake replied:
I think this is fine. We are in alpha, the existence of an unused role in the databases of those that are testing the nightlies is not going to cause any harm. And yes the first library clean will remove them.

Montellese commented:
I'm still not sure about the `allroles` parameter, especially on `GetAlbums` and `GetSongs` where I want to retrieve albums and songs and a parameter named `allroles` just feels wrong.

Tolriq commented:
@DaveTBlake I always comment on things related to PR, the only things not related are important things discovered during testing, like the Windows 10 port 80 bugs or the reboot during install.
For the first one despite Trac + Forum I had to fix myself 7 month later to have the fix in Jarvis.
For the reboot it will be the same :)

About my remarks here, like the last one about translations, as you seems to not understand what I write since as always my English suxx, You did say if there's a reason found for fixed ID, I was just giving one .... 

> but it is not a crisis.

I never said so and multiple times explained that is was not so important, but if no one report problems or start fixing then it's never done.

And you say to cut of a debate, but there was no debate in this thread, no blocking, nothing from my side.

About API sub optimal, it's because I see way larger that Kodi GUI. Even if I see long term, but all those hiding things from API is what limits external tools like Ember media Manager or even Plex / Emby addons to better integrate with Kodi.

I do not only preach for Yatse or remotes but for the global picture. If it was only me I would never have fixed the EventServer security problem as it will trigger mega tons of support on my side ;)

About GUI acting normally, I first had make tests with a compilation that have the standard DJ-MIx role (auto set from MusicBrainz) and never see it, leading to the question.

DaveTBlake commented:
>I'm still not sure about the `allroles` parameter, especially on GetAlbums and GetSongs where I want to retrieve albums and songs and a parameter named allroles just feels wrong.

@Montellese both methods have artist as a filter, so that the API can filter on just original artist role (as before) but also filter by that artist regardless of role a parameter is needed. I can't think of a better name.

Perhaps an example will help. Say Dolly Parton, the country singer, is the album artist on  number of albums in your library. She has also composed or wrote the lyrics on a number of songs on other people's albums. The API  (and GUI) originally would just fetch the songs or albums performed by Dolly Parton, that is the main use and we need to keep it as default. But Kodi now knows those songs, and the albums that contain them, that she composed, so the API can return those when role=composer filter specified. Finally Kodi can return the songs and albums where Dolly played any role, performed, composed, wrote lyrics or sang backing vocals etc. In the spirit of all data being available via API it needs a way to specify that the songs with any relationship to Dolly are returned.

Originally I had a "magic number" for role meaning all roles, we switched to a parameter rather than that. I see "allroles" as a better solution, and I don't see what else to do.

DaveTBlake committed:
Name id, genre pair generically for use other than songgenre.
Correct "allsongs" parameter default and description.
Space filter parameter layout for readability.

DaveTBlake commented:
Have made small changes @Montellese suggested, will squish before merge.
@joethefox you made need to note the renaming of `Audio.Artist.SongGenres` and elements to `Audio.Details.Genres` with id and name.

@Montellese if you can convince yourself that "allroles" is acceptable as a parameter then I hope this is good to merge.

Tolriq commented:
@DaveTBlake me again sorry, but is it possible to be consistent with Video ? (https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/types.json#L399) genreid and title ?

DaveTBlake commented:
>@DaveTBlake me again sorry, but is it possible to be consistent with Video ? genreid and title ?

OK. `Audio.Artist.SongGenres` and becomes  `Audio.Details.Genres with genreid and title.

Montellese commented:
@Tolriq: It will not be completely compatible because the video object also has `label` and `thumbnail`.

DaveTBlake commented:
It is an array of values too, not just a single pair. I looked at just using the video item and decided it was not quite right. But using same names for the elements makes sense.

Tolriq commented:
Yes it's purely about naming convention.

It's like duration / runtime for exemple. It's easier to use the same field name for the same kind of value that avoid reading the introspect all the time.

Montellese commented:
@DaveTBlake: we could allow the value `all` to be passed to a `role` filter. That wouldn't work with `roleid` but `role` is available as an alternative to `roleid` in all filters.

Tolriq commented:
@Montellese this would introduce the concept of adding a filter to unfilter a list. Does not seems logical and clean for an API :(

DaveTBlake commented:
@Montellese filter "role" = "allroles" was what I offered back in January. In discussion with you it was decided to go with a separate parameter instead, much for the reasons @Tolriq suggests.

I feel I am going in circles. :(
I will do whatever you want, but can we please just get this completed.

Montellese commented:
I'm in conflict with myself about this. Having an `allroles` parameter on `AudioLibrary.GetSongs` just looks ugly and not logical. The fact that it then even has the possibility to override filters (which should always apply to the list that has been parametrized with the rest of the parameters) makes it even worse. But then again ideally not providing any filters should give you the whole list which in this case simply won't be possible if we want to keep backwards compatibility.

So IMO we have these options

1. drop backwards compatibility and return all artists independent of their role(s)
2. use the `allroles` parameter which to me is ok on `GetArtists` but not at all intuitive on `GetAlbums` and `GetSongs` and has the possibility to override other parameters which isn't intuitive either
3. allow `all` and `artist` (or whatever) as special values for the `role` filters with `artist` being the default

Tolriq commented:
3) is adding an implicit default filter that needs to be added to be removed, seems as (or more) ugly than allroles.

If it was only me I'd vote 1. 

But about 2 maybe it's just a naming problem that can be solved ?
Do something like norolefilter sounds better for albums / songs ? Maybe it makes it's role more clear.

DaveTBlake commented:
1) is unnecessarily brutal. I really don't want my new functionality to be a hated addition that caused API users a load of work. This PR is only a breaking change because of the rename of GetContributors to GetRoles, for most API consumers they will not have been impacted at all.

I feel whatever we do it should be consistent for GetArtists, GetAlbums and GetSongs. The current implementation is 2), having replaced 3) that I had previously suggested. I now see 2) as better than 3).

But if you want 3) then I will change things back, but I believe it should be like that in all 3 methods.

DaveTBlake commented:
@Montellese sorry that you are conflicted, but can you please either approve this implementation or decide how you  want it changed so that I can complete it.

I am happy with currently implemented 2), @Tolriq is reluctantly accepting of it, @joethefox is using this work already. But if you want 1) or 3) (applied to all 3 GetXXXX methods) then I will do that. It helps no one for this to drag on any further.

DaveTBlake commented:
@Montellese what can I do to move this forwards please? Can it be merged?

DaveTBlake commented:
jenkins build this please
